spring-ai-core/src/main/java/org/springframework/ai/model/function/MethodFunctionCallback.java

```diff
@@ -0,0 +1,269 @@
+
+	/**
+	 * The JSON schema generated from the method input parameters.
+	 */
+	private final String inptuSchema;
+
+	public MethodFunctionCallback(Object functionObject, Method method, String description, ObjectMapper mapper) {
+
+		Assert.notNull(method, "Method must not be null");
+		Assert.notNull(mapper, "ObjectMapper must not be null");
+		Assert.hasText(description, "Description must not be empty");
+
+		this.method = method;
+		this.description = description;
+		this.mapper = mapper;
+		this.functionObject = functionObject;
+
+		Assert.isTrue(this.functionObject != null || Modifier.isStatic(this.method.getModifiers()),
+				"Function object must be provided for non-static methods!");
+
+		// Generate the JSON schema from the method input parameters
+		Map<String, Class<?>> methodParameters = Stream.of(method.getParameters())
+			.collect(Collectors.toMap(param -> param.getName(), param -> param.getType()));
+		this.inptuSchema = generateJsonSchema(methodParameters);
+
+		logger.info("Generated JSON Schema: \n:" + this.inptuSchema);
+	}
+
+	@Override
+	public String getName() {
+		return method.getName();
+	}
+
+	@Override
+	public String getDescription() {
+		return this.description;
+	}
+
+	@Override
+	public String getInputTypeSchema() {
+		return this.inptuSchema;
+	}
+
+	@Override
+	public String call(String functionInput) {
+
+		try {
+
+			Map<String, Object> map = this.mapper.readValue(functionInput, Map.class);
+
+			Object[] methodArgs = Stream.of(this.method.getParameters()).map(parameter -> {
+				Object rawValue = map.get(parameter.getName());
+				Class<?> type = parameter.getType();
+				return this.toJavaType(rawValue, type);
+			}).toArray();
+
+			Object response = this.method.invoke(this.functionObject, methodArgs);
+
+			var returnType = this.method.getReturnType();
+			if (returnType == Void.TYPE) {
+				return "Done";
+			}
+
+			if (returnType == Class.class || returnType.isRecord() || returnType == List.class
+					|| returnType == Map.class) {
+				return ModelOptionsUtils.toJsonString(response);
+
+			}
+			return "" + response;
+		}
+		catch (Exception e) {
+			throw new RuntimeException(e);
+		}
+
+	}
+
+	/**
+	 * Generates a JSON schema from the given named classes.
+	 * @param namedClasses The named classes to generate the schema from.
+	 * @return The generated JSON schema.
+	 */
+	protected String generateJsonSchema(Map<String, Class<?>> namedClasses) {
+		try {
+			JsonSchemaGenerator schemaGen = new JsonSchemaGenerator(this.mapper);
+
+			ObjectNode rootNode = this.mapper.createObjectNode();
+			rootNode.put("$schema", "https://json-schema.org/draft/2020-12/schema");
+			rootNode.put("type", "object");
+			ObjectNode propertiesNode = rootNode.putObject("properties");
+
+			for (Map.Entry<String, Class<?>> entry : namedClasses.entrySet()) {
+				String className = entry.getKey();
+				Class<?> clazz = entry.getValue();
+
+				JsonSchema schema = schemaGen.generateSchema(clazz);
```

Comment:
looking into `BeanOutputConverter` we could use the following code to reuse the "victools" json schema that I believe is what supports the 2020_12 draft as standard jackson doesn't support that version of the schema.  If jackson now supports it, that would be great.


```
		JacksonModule jacksonModule = new JacksonModule();
		SchemaGeneratorConfigBuilder configBuilder = new SchemaGeneratorConfigBuilder(DRAFT_2020_12, PLAIN_JSON)
			.with(jacksonModule);
		SchemaGeneratorConfig config = configBuilder.build();
		SchemaGenerator generator = new SchemaGenerator(config);
		JsonNode jsonNode = generator.generateSchema(this.typeRef.getType());
        ...  convert node to string..


```